src/ir/mod.rs

```diff
@@ -747,13 +790,19 @@ impl<T> UCanonical<T> {
 }
 
 impl UCanonical<InEnvironment<Goal>> {
-    /// A goal has coinductive semantics if it is of the form `T: AutoTrait`.
+    /// A goal has coinductive semantics if it is of the form `T: AutoTrait`, or if it is of the
+    /// form `WellFormed(T: Trait)` where `Trait` is any trait. The latter is needed for dealing
```

Comment:
I am mildly surprised that we don't want to just make *all* `WellFormed` goals coinductive. But I guess it's only needed for traits...

Reply:
Well types only have where clauses of the form `T: Trait`, there are no `WF(Type)` where clauses so you never have the kinds of cycles encountered with traits.